hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -8,9 +8,449 @@ import is.hail.expr.types.physical._
+        jp.define(_ => terminatingStream.load.ceq(i).mux(
+          next,
+          Code(closes(i), next)))
+      }
+
+      source.copy(setup0 = Code(firstPull := true, source.setup0), setup = Code(firstPull := true, source.setup))
+    }
+  }
+
+  def mux[A: ParameterPack](cond: Code[Boolean], left: Stream[A], right: Stream[A]): Stream[A] = new Stream[A] {
+    def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+      val b = newLocal[Code[Boolean]]
+      val eosJP = joinPoint()
+      val pushJP = joinPoint[A]
+
+      eosJP.define(_ => eos)
+      pushJP.define(push)
+
+      val l = left(eosJP(()), pushJP(_))
+      val r = right(eosJP(()), pushJP(_))
+
+      val lPullJP = joinPoint()
+      val rPullJP = joinPoint()
+
+      lPullJP.define(_ => l.pull)
+      rPullJP.define(_ => r.pull)
+      Source[A](
+        setup0 = Code(b := cond, l.setup0, r.setup0),
+        close0 = Code(l.close0, r.close0),
+        setup = b.load.mux(l.setup, r.setup),
+        close = b.load.mux(l.close, r.close),
+        firstPull = (l.firstPull, r.firstPull) match {
+          case (None, None) => None
+          case _ => Some(b.load.mux(l.firstPull.getOrElse(lPullJP(())), r.firstPull.getOrElse(rPullJP(()))))
+        },
+        pull = JoinPoint.mux(b.load, lPullJP, rPullJP)
+      )
+    }
+  }
+
+  def fromParameterized[P, A](
+    stream: EmitStream.Parameterized[P, A]
+  ): P => COption[Stream[A]] = p => new COption[Stream[A]] {
+    def apply(none: Code[Ctrl], some: Stream[A] => Code[Ctrl])(implicit ctx: EmitStreamContext): Code[Ctrl] = {
+      import EmitStream.{Missing, Start, EOS, Yield}
+      implicit val sP = stream.stateP
+      val s = newLocal[stream.S]
+      val sNew = newLocal[stream.S]
+
+      def src(s0: stream.S): Stream[A] = new Stream[A] {
+        def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+          Source[A](
+            setup0 = Code(s.init, sNew.init),
+            close0 = Code._empty,
+            setup = sNew := s0,
+            close = Code._empty,
+            firstPull = None,
+            pull = Code(s := sNew.load, stream.step(s.load) {
+              case EOS => eos
+              case Yield(elt, s1) => Code(sNew := s1, push(elt))
+            }))
+        }
+      }
+
+      stream.init(p) {
+        case Missing => none
+        case Start(s0) => some(src(s0))
+      }
+    }
+  }
+}
+
+object EmitStream2 {
+  import CodeStream._
+
+  private[ir] def apply(
+    emitter: Emit,
+    streamIR0: IR,
+    env0: Emit.E,
+    er: EmitRegion,
+    container: Option[AggContainer]
+  ): COption[Stream[EmitTriplet]] = {
+    val fb = emitter.mb.fb
+
+    def emitIR(ir: IR, env: Emit.E): EmitTriplet =
+      emitter.emit(ir, env, er, container)
+
+    def emitStream(streamIR: IR, env: Emit.E): COption[Stream[EmitTriplet]] =
+      streamIR match {
+
+        case _ =>
```

Comment:
?

Reply:
This is where I will add emit implementations for the other stream nodes. Right now there is just the default case, which uses the old `EmitStream`, and wraps the result in a conversion to the new `Stream`. That way I could test new emit cases one at a time.